flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SinkTransformationTranslator.java

```diff
@@ -0,0 +1,283 @@
+import org.apache.flink.streaming.api.transformations.SinkTransformation;
+import org.apache.flink.streaming.runtime.operators.sink.BatchCommitterOperatorFactory;
+import org.apache.flink.streaming.runtime.operators.sink.BatchGlobalCommitterOperatorFactory;
+import org.apache.flink.streaming.runtime.operators.sink.GlobalStreamingCommitterOperatorFactory;
+import org.apache.flink.streaming.runtime.operators.sink.StatefulWriterOperatorFactory;
+import org.apache.flink.streaming.runtime.operators.sink.StatelessWriterOperatorFactory;
+import org.apache.flink.streaming.runtime.operators.sink.StreamingCommitterOperatorFactory;
+import org.apache.flink.streaming.util.graph.StreamGraphUtils;
+
+import java.util.Collection;
+import java.util.Collections;
+
+/**
+ * A {@link TransformationTranslator} for the {@link SinkTransformation}.
+ */
+@Internal
+public class SinkTransformationTranslator<InputT, CommT, WriterStateT, GlobalCommT> implements
+		TransformationTranslator<Object, SinkTransformation<InputT, CommT, WriterStateT, GlobalCommT>> {
+
+	@Override
+	public Collection<Integer> translateForBatch(
+			SinkTransformation<InputT, CommT, WriterStateT, GlobalCommT> transformation,
+			Context context) {
+
+		StreamGraphUtils.validateTransformationUid(context.getStreamGraph(), transformation);
+
+		int writerId = addWriter(transformation, context);
+		int committerId = addCommitter(
+				writerId,
+				transformation,
+				new BatchCommitterOperatorFactory<>(transformation.getSink()),
+				context);
+		addGlobalCommitter(
+				committerId >= 0 ? committerId : writerId,
+				transformation,
+				new BatchGlobalCommitterOperatorFactory<>(transformation.getSink()),
+				context);
+		return Collections.emptyList();
+	}
+
+	@Override
+	public Collection<Integer> translateForStreaming(
+			SinkTransformation<InputT, CommT, WriterStateT, GlobalCommT> transformation,
+			Context context) {
+
+		StreamGraphUtils.validateTransformationUid(context.getStreamGraph(), transformation);
+
+		int writerId = addWriter(transformation, context);
+		int committerId = addCommitter(
+				writerId,
+				transformation,
+				new StreamingCommitterOperatorFactory<>(transformation.getSink()),
+				context);
+		addGlobalCommitter(
+				committerId >= 0 ? committerId : writerId,
+				transformation,
+				new GlobalStreamingCommitterOperatorFactory<>(transformation.getSink()),
+				context);
+
+		return Collections.emptyList();
+	}
+
+	/**
+	 * Add a sink writer node to the stream graph.
+	 *
+	 * @param sinkTransformation The transformation that the writer belongs to
+	 *
+	 * @return The stream node id of the writer
+	 */
+	private int addWriter(
+			SinkTransformation<InputT, CommT, WriterStateT, GlobalCommT> sinkTransformation,
+			Context context) {
+		final boolean hasState = sinkTransformation
+				.getSink()
+				.getWriterStateSerializer()
+				.isPresent();
+		final TypeInformation<CommT> outTypeInfo = TypeExtractor.createTypeInfo(
```

Comment:
I guess this will change after https://github.com/apache/flink/pull/13825 but we could even move it to a separate method as it is used in many places.

Reply:
yes.